### ISSUE TITLE
Mismatches between log message formats and parameters

### ISSUE THREAD

**Issue by algr** [DynamoRIO/dynamorio#1893]
If we assume that the LOG messages take 'standard' printf formats, i.e. the format string doesn't have any custom processing, then we can use GCC's function attribute for printf format checking, as in the diff below (this ought to be guarded for Windows and non-GCC compilers, this is just to show how it's done).

When this checking is enabled, there are hundreds of warnings and errors from the debug build.

Most diagnostics are because pointers are passed to PFX formats where PFX is defined as e.g. "%016lx".  This will probably work at runtime, but really these formats would be better as "%016p".

A few diagnostics are more serious and show mismatch in size for integer parameters, e.g. size_t being passed to "%d" (see e.g. core/utils.c line 3464).  These formats should be "%zd".  globals_shared.h already defines a SZFMT but a lot of messages don't use it.

Even more seriously there are mismatches in number of arguments, e.g. this from core/link.c where the format expects 6 arguments and only 5 are passed:
{code}
                     LOG(THREAD, LOG_LINKS, 4, "    not linking F%d("PFX")."PFX" -> "
                         "F%d("PFX" == "PFX") (self-link already linked)\n",
                         in_f->id, in_f->tag, EXIT_CTI_PC(in_f, l),
                         new_f->id, new_f->tag);
{code}

Here's the change to enable the GCC format checking:
{code}
@@ -1166,13 +1166,14 @@ bool bitmap_check_consistency(bitmap_t b, uint bitmap_size, uint expect_free);
 # define LOG_DECLARE(declaration)
 # define DOCHECK(level, statement) /* nothing */
 #endif
-void print_log(file_t logfile, uint mask, uint level, const char *fmt, ...);
-void print_file(file_t f, const char *fmt, ...);
+
+void print_log(file_t logfile, uint mask, uint level, const char *fmt, ...) __attribute__((format(printf,4,5)));
+void print_file(file_t f, const char *fmt, ...) __attribute__((format(printf,2,3)));
 
 /* For repeated appending to a buffer.  The "sofar" var should be set
  * to 0 by the caller before the first call to print_to_buffer.
  */
-bool print_to_buffer(char *buf, size_t bufsz, size_t *sofar INOUT, const char *fmt, ...);
+bool print_to_buffer(char *buf, size_t bufsz, size_t *sofar INOUT, const char *fmt, ...) __attribute__((format(printf,4,5)));
{code}

Before I launch on writing a patch that fixes all the cases, can I have your feedback on whether you would like this format checking to be enabled?

**Comment by derekbruening** [comment]
We rejected %p long ago due to inconsistency in how it is handled on Windows vs Linux.  Please only consider formatting that is portable.

Today we use our own implementation of printf that supports common codes but not all codes.

**Comment by algr** [comment]
But if you use your own implementation of printf then you can make %p behave consistently across Windows and Linux.

The GCC attributes don't force you to use system printf (with any inconsistency that implies), they just check arguments against the format.  And clearly that's picking up some cases where the log messages would be corrupt if they would ever trigger.  If you've got log messages that print undefined data, I think that's an even bigger issue than log messages that print pointers a different way on Windows vs. Linux, surely?

**Comment by derekbruening** [comment]
Yes, the size mismatches and number of arg mismatches should clearly be fixed, and this sounds like a good check in general *if the signal to noise ratio can be controlled*.

Please do not swap which format codes are used for the size-varying pointer-and-integer format macros without a discussion and consideration: they are the way they are today for a reason.  Originally our own printf code was only used on Linux, with ntdll's printf used on Windows.  Although we have since swapped to using our own printf on Windows as well for the core, this is not the case for standalone apps or apps using the drdecode library.  Swapping to %p may well cause those tests to fail.  Using %p for "golden output" tests is simply fragile.  The ideal situation would be to be able to suppress warnings like pointer vs integer types while keeping the other diagnostics.

**Comment by algr** [comment]
Here's a proposal:

   - define one set of formats based on "%p" for use in debug-mode diagnostic LOG messages.  Checking against golden output tests is not an issue.  Pointer arguments to LOG can be passed as is.

   - define another set of formats based on "%lx" (etc.) for use in user messages.  Pointer arguments to user messages will be wrapped in a macro e.g. PTR(x) which will cast the pointer to an integer type matching the format

**Comment by derekbruening** [comment]
Re-visiting: probably switching to %p is the best thing to do.  Moving to VS2017 (#2166) also hits a lot of format string warnings which would be best to not have to suppress.  Going to try it.

**Comment by derekbruening** [comment]
The switch to %p is in place.  @algr please go ahead with your proposal to enable more format string warnings and fix revealed bugs.

**Comment by algrant-arm** [comment]
I am picking up a few places where PFX is being used to print a reg_t. I could change it to PIFX but it would be better to have a REGFX in case reg_t isn't the size of a native pointer.

Defining types and formats using stdint.h and inttypes.h would simplify the code a lot - and stdint.h at least is already used in some places. Is it now acceptable to have a dependency on these headers throughout the codebase? Or does it still need to work with very old versions of Visual Studio?

I think at least it would make the code more maintainable if we adopted the same convention as inttypes.h, i.e. format macros are of the form PRI<base><type> and do not include the "%".

**Comment by derekbruening** [comment]
Visual Studio 2013 update 5 is the currently supported version, but we are in the process of updating to 2017 (#3860) and will no longer actively support 2013 once Appveyor is switched.  Update 5 I think is pretty close to full C99 (finally allows var decls after an if(), unlike update 4).  It must have stdint.h since those parts of the code base build there.

**Comment by algrant-arm** [comment]
I'm working my way through issues, and there are a few that regularly occur and are easy to fix:

- printing a ptr_uint_t with PFX (should be PIFX)
- using a size_t as a field width (should be cast to int)

But I'm struggling with some of the logging and reporting macros which end up as d_r_notify. E.g. SYSLOG and FATAL_USAGE_ERROR. It's unclear which macros take a "format plus args" sequence. Is there some documentation on how to use the message macros, i.e. which macro would be used in which situation and what arguments should be passed?

**Comment by derekbruening** [comment]
> * printing a ptr_uint_t with PFX (should be PIFX)

I suspect that these uses cases want leading zeroes, so perhaps a new code should be made for them to preserve the zeroes.

> * using a size_t as a field width (should be cast to int)

Like #3947?

> But I'm struggling with some of the logging and reporting macros which end up as d_r_notify. E.g. SYSLOG and FATAL_USAGE_ERROR. It's unclear which macros take a "format plus args" sequence. Is there some documentation on how to use the message macros, i.e. which macro would be used in which situation and what arguments should be passed?

Those use the special formatting strings in core/win32/events.mc which go through the Windows message compiler (for the event log) and on UNIX go through a printf-ish routine.
